c++/src/kj/mutex.h

```diff
@@ -172,6 +220,44 @@ class Mutex {
   mutable pthread_rwlock_t mutex;
 #endif
 
+#if KJ_SAVE_ACQUIRED_LOCK_INFO
+  uint32_t lockedExclusivelyByThread = 0;
+  SourceLocation lockAcquiredLocation;
+
+  volatile bool updatedSharedMetadata = false;
+
+  void acquiredExclusive(uint32_t tid, const SourceLocation& location) noexcept {
+    __atomic_store_n(&lockedExclusivelyByThread, tid, __ATOMIC_RELEASE);
```

Comment:
My understanding is that the rough mapping for atomics is that read => acquire, write => release.Since this is a write operation, isn't release the right operation?

Reply:
No, load vs. store and acquire vs. release are separate things. Acquire-load and release-store are the two combinations that are "free" on x86, but that doesn't meant all loads should be acquires and all releases stores.

Acquire vs. release are about whether the "protected" data access happens before or after this operation. In this case, you perform this store before accessing the locked data. So if anything, it's an acquire.

That said, since this particular flag is used only for debugging, it might be worth using `__ATOMIC_RELAXED` instead, to avoid invoking the processor's locking protocol, which is expensive. A relaxed store is just a regular store, whereas an acquire-store requires instructing the processor to perform a locking instruction. The down side of using relaxed is that it might take longer for another core to pick up this change, but that's probably not a big deal in practice because we don't start trying to trace locks until they are held for a long time. And anyway, since you're not actually setting this value atomically at the same time as taking the lock, such a race condition exists either way.

So I'd go with `__ATOMIC_RELAXED` here.

For `releasingExclusive()` below, a release-store is "free", but I'd still probably change it to `__ATOMIC_RELAXED` for consistency.

Reply:
I don't think acquire semantics have any meaning for stores, nor does release semantics have any meaning for a load.

The [documentation](https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html) for `__atomic_store_n` says: The valid memory order variants are __ATOMIC_RELAXED, __ATOMIC_SEQ_CST, and __ATOMIC_RELEASE. There are similar rules for c++'s  `std::atomic_store_explicit` and `std::atomic<T>::store` (which specifies undefined behavior if it's not one of those three memory orderings).

Reply:
@Dr-Emann That's odd that the docs say that, when the compiler does in fact implement unique behavior for acquire-store that is different from any of the models that the document says are valid.

That said, it's true that acquire semantics on a pure-store operation is almost never useful. This is not because they inherently don't make sense, but rather because there's no way for another thread to correctly synchronize with it. The other thread would need some sort of "compare-and-store" operation, where it can atomically check the value of one location and, based on its value, conditionally store to another location. Generally, CPUs don't implement atomic ops that transact over multiple locations at once, as that would be incredibly complicated. But if they did, then acquire-store could theoretically make sense.

(This point is essentially the same point as I made above when I pointed out that using acquire semantics here wouldn't really accomplish much as the potential race condition still exists.)

However, that doesn't mean that "stores should always have release semantics". Acquire semantics make sense on atomic operations that are both load-and-store (e.g. atomic increment/decrement, compare-and-swap, exchange, etc.) -- and that's not just because "acquires are for loads", it's usually important in these cases that the "store" part has acquire semantics as well.

ANYWAY

Release semantics don't make sense here, because nothing is being released. Acquire semantics don't make sense, because they don't really solve any particular race condition. So relaxed is the way to go.

PS. Out of curiosity, @Dr-Emann, who are you and why do you read all our PR threads?

Reply:
I found capnproto looking for an RPC framework for work, and it was the only RPC system that really fit what I wanted (bidirectional data, especially typed "server" initiated RPC calls). I ended up getting pretty into it, it's a really elegant system. I set myself as a watcher for issues/PRs just to keep up with new features, in case there were any which looked so nice I'd upgrade before the next release, or what to look forward to in the next release (really excited for cross thread fulfillers and addRef on kj::Executor).

I hope you don't mind.